File: lesson_5/src/test/java/ru/academits/OnlineStore.java
package ru.academits;

import io.github.bonigarcia.wdm.WebDriverManager;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.openqa.selenium.*;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.edge.EdgeDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.Select;


public class OnlineStore {

    private WebDriver driver;

    @BeforeEach
    public void setUp() {

        String browser = System.getProperty("browser");

        if (browser.equals("chrome")) {
            WebDriverManager.chromedriver().setup();
            driver = new ChromeDriver();
        } else if (browser.equals("firefox")) {
            WebDriverManager.firefoxdriver().setup();
            driver = new FirefoxDriver();
        } else if (browser.equals("edge")) {
            WebDriverManager.edgedriver().setup();
            driver = new EdgeDriver();
        }

        String url = "https://demowebshop.tricentis.com/";

        driver.get(url);
        driver.manage().window().maximize();

        System.out.println("Current url = " + driver.getCurrentUrl());
        Assertions.assertEquals(url, driver.getCurrentUrl());
    }


    @AfterEach
    public void tearDown() {
        if (driver != null) {
            driver.quit();
        }
    }


    @ParameterizedTest
    @ValueSource(strings = {"Laptop", "Smartphone", "Fiction"})
    public void testAddToCart  (String searchQuery) {
        // Ввод поискового запроса
        WebElement searchBox = driver.findElement(By.xpath("//*[@id='small-searchterms']"));
        searchBox.sendKeys(searchQuery);
        searchBox.sendKeys(Keys.ENTER);

        // Нажатие кнопки "Add to cart" для первого товара и сохранение наименования товара
        WebElement addToCart = driver.findElement(By.xpath("//*[@value='Add to cart'][1]"));

        String expectedItem = driver.findElement(By.xpath("//*[@class='item-box'][1]//h2/a")).getText();
        addToCart.click();

        // Переход в корзину
        WebElement cart = driver.findElement(By.xpath("//*[contains(text(), 'Shopping cart')]"));
        cart.click();

        // Проверка наличия в корзине добавленного товара
        WebElement elem = driver.findElement(By.xpath("//tbody/tr[1]//a"));
        Assertions.assertEquals(expectedItem, elem.getText(), "Товар не найден в корзине");
    }


    @Test
    public void testProductDisplay() {
        // Открыть раздел "Books"
        driver.findElement(By.xpath("//*[@class='top-menu']//a[contains(text(),'Books')]")).click();

        // Проверить отображение 8 товаров
        selectItemsPerPage("8");
        verifyItemsDisplayed(8);

        // Проверить отображение 4 товаров
        selectItemsPerPage("4");
        verifyItemsDisplayed(4);
    }


    private void selectItemsPerPage(String value) {
        WebElement displayDropdown = driver.findElement(By.xpath("//*[@id='products-pagesize']"));
        displayDropdown.click();
        Select select = new Select(displayDropdown);
        select.selectByVisibleText(value);
    }


    private void verifyItemsDisplayed(int expectedCount) {
        int actualCount = driver.findElements(By.xpath("//*[@class='product-grid']/div")).size();

        Assertions.assertTrue(actualCount <= expectedCount,
                "Отображается " + actualCount + " товаров, хотя должно быть не более " + expectedCount);

        System.out.println("При отображении " + expectedCount + " товаров на странице, найдено: " + actualCount);
    }
}

File: lesson_5_python/online_store.py
import pytest
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from webdriver_manager.chrome import ChromeDriverManager
from webdriver_manager.firefox import GeckoDriverManager
from webdriver_manager.microsoft import EdgeChromiumDriverManager
import time

class TestOnlineStore:
    @pytest.fixture(autouse=True)
    def setup(self, request):
        # Получаем параметр браузера из командной строки
        browser = request.config.getoption("--browser")

        # Инициализация драйвера в зависимости от браузера
        if browser == "chrome":
            self.driver = webdriver.Chrome(ChromeDriverManager().install())
        elif browser == "firefox":
            self.driver = webdriver.Firefox(executable_path=GeckoDriverManager().install())
        elif browser == "edge":
            self.driver = webdriver.Edge(EdgeChromiumDriverManager().install())
        else:
            raise ValueError(f"Unsupported browser: {browser}")

        self.url = "https://demowebshop.tricentis.com/"
        self.driver.get(self.url)
        self.driver.maximize_window()

        print(f"Current URL = {self.driver.current_url}")
        assert self.driver.current_url == self.url

        yield

        # Закрытие браузера после теста
        if self.driver:
            self.driver.quit()

    @pytest.mark.parametrize("search_query", ["Laptop", "Smartphone", "Fiction"])
    def test_add_to_cart(self, search_query):
        # Ввод поискового запроса
        search_box = WebDriverWait(self.driver, 10).until(
            EC.presence_of_element_located((By.ID, "small-searchterms"))
        )
        search_box.send_keys(search_query)
        search_box.send_keys(Keys.ENTER)

        # Нажатие кнопки "Add to cart" для первого товара
        add_to_cart = WebDriverWait(self.driver, 10).until(
            EC.presence_of_element_located((By.XPATH, "(//input[@value='Add to cart'])[1]"))
        )

        # Сохранение названия товара
        expected_item = WebDriverWait(self.driver, 10).until(
            EC.presence_of_element_located((By.XPATH, "(//div[@class='item-box'])[1]//h2/a"))
        ).text

        # Прокрутка и клик
        self.driver.execute_script("arguments[0].scrollIntoView();", add_to_cart)
        add_to_cart.click()

        # Ожидание сообщения об успешном добавлении
        WebDriverWait(self.driver, 10).until(
            EC.visibility_of_element_located((By.CSS_SELECTOR, ".bar-notification.success"))
        )

        # Переход в корзину
        cart = WebDriverWait(self.driver, 10).until(
            EC.element_to_be_clickable((By.XPATH, "//a[contains(text(), 'Shopping cart')]"))
        )
        cart.click()

        # Проверка наличия товара в корзине
        actual_item = WebDriverWait(self.driver, 10).until(
            EC.presence_of_element_located((By.XPATH, "//tbody/tr[1]//a"))
        ).text

        assert actual_item == expected_item, f"Товар {expected_item} не найден в корзине"

def pytest_addoption(parser):
    parser.addoption(
        "--browser", action="store", default="chrome", help="browser to execute tests (chrome, firefox, edge)"
    )